lib/backend/firestore/firestorebk.go

```diff
@@ -0,0 +1,716 @@
+		r.Expires = replaceWith.Expires.UTC().Unix()
+	}
+
+	_, err = expectedDocSnap.Ref.Set(ctx, r)
+	if err != nil {
+		return nil, ConvertGRPCError(err)
+	} else {
+		return b.newLease(replaceWith), nil
+	}
+}
+
+// Delete deletes item by key
+func (b *FirestoreBackend) Delete(ctx context.Context, key []byte) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	docRef := b.svc.Collection(b.CollectionName).Doc(b.hashKey(key))
+	doc, _ := docRef.Get(ctx)
+
+	if !doc.Exists() {
+		return trace.NotFound("key %s does not exist", string(key))
+	} else {
+		_, err := docRef.Delete(ctx)
+
+		if err != nil {
+			return ConvertGRPCError(err)
+		} else {
+			return nil
+		}
+	}
+}
+
+// NewWatcher returns a new event watcher
+func (b *FirestoreBackend) NewWatcher(ctx context.Context, watch backend.Watch) (backend.Watcher, error) {
+	select {
+	case <-b.watchStarted.Done():
+	case <-ctx.Done():
+		return nil, trace.ConnectionProblem(ctx.Err(), "context is closing")
+	}
+	return b.buf.NewWatcher(ctx, watch)
+}
+
+// KeepAlive keeps object from expiring, updates lease on the existing object,
+// expires contains the new expiry to set on the lease,
+// some backends may ignore expires based on the implementation
+// in case if the lease managed server side
+func (b *FirestoreBackend) KeepAlive(ctx context.Context, lease backend.Lease, expires time.Time) error {
+	if len(lease.Key) == 0 {
+		return trace.BadParameter("lease is missing key")
+	}
+	docSnap, err := b.svc.Collection(b.CollectionName).Doc(b.hashKey(lease.Key)).Get(ctx)
+	if err != nil {
+		return ConvertGRPCError(err)
+	}
+
+	if !docSnap.Exists() {
+		return trace.NotFound("key %s does not exist, cannot extend lease", lease.Key)
+	}
+
+	var r record
+	_ = docSnap.DataTo(&r)
+
+	if r.isExpired() {
+		return trace.NotFound("key %s has already expired, cannot extend lease", lease.Key)
+	}
+
+	updates := []firestore.Update{{Path: expiresDocProperty, Value: expires.UTC().Unix()}, {Path: timestampDocProperty, Value: b.clock.Now().UTC().Unix()}}
+	_, err = docSnap.Ref.Update(ctx, updates)
+	if err != nil {
+		return ConvertGRPCError(err)
+	}
+	return err
+}
+
+// Close closes the Firestore client contexts and releases associated resources
+func (b *FirestoreBackend) Close() error {
+	b.clientCancel()
+	err := b.buf.Close()
+	if err != nil {
+		b.Logger.Error("error closing buffer, continuing with closure of other resources...", err)
+	}
+	return b.svc.Close()
+}
+
+// CloseWatchers closes all the watchers without closing the backend
+func (b *FirestoreBackend) CloseWatchers() {
+	b.buf.Reset()
+}
+
+// Clock returns wall clock
+func (b *FirestoreBackend) Clock() clockwork.Clock {
+	return b.clock
+}
+
+// newLease
+func (b *FirestoreBackend) newLease(item backend.Item) *backend.Lease {
+	var lease backend.Lease
+	if item.Expires.IsZero() {
+		return &lease
+	}
+	lease.Key = item.Key
+	return &lease
+}
+
+// hashKey hashes keys using SHA1 enabling fast lookups
```

Comment:
Let's use reversible encoding here, instead of sha1 - e.g. hex or base64 depending on what works for  firebase, alternatively, we can use replacement escaping symbol for `\`

Reply:
Going w/ string replacement so that document IDs are easy to find in the Firestore UI.